File: 0053-maximum-subarray/0053-maximum-subarray.js
/**
 * @param {number[]} nums
 * @return {number}
 */

var maxSubArray = function(nums) {

    // initialize maxSum and currSum:
    let maxSum = nums[0];
    let currSum = 0;
    
    for (let num of nums) {
        
        if (currSum < 0) {currSum = 0}
        
        currSum = currSum + num;
        maxSum = Math.max(maxSum, currSum);
    }
    
    return maxSum;
};

File: 0053-maximum-subarray/0053-maximum-subarray.java
// linear time sln: loop through the array; keep track of 2 variables: maxSub, currSum
// greedy: when currSum < 0, reset to 0 and continue

// subArray: must be continuous

class Solution {
    public int maxSubArray(int[] nums) {
        
        int maxSub = nums[0];
        int currSum = 0;
        
        // loop through nums:
        for (int number:nums) {
            
            // if currSum < 0, then reset to 0
            if (currSum < 0) {
                
                currSum = 0;
            }
            
            currSum = currSum + number;
            maxSub = Math.max(maxSub, currSum);
        }
        
        return maxSub;
    }
}